pkg/query-service/app/preferences/model.go

```diff
@@ -0,0 +1,479 @@
+	isPreferenceEnabledAtUserScope := preference.IsEnabledForScope(UserAllowedScope)
+	if !isPreferenceEnabledAtUserScope {
+		return nil, &model.ApiError{Typ: model.ErrorForbidden, Err: fmt.Errorf("preference is not enabled at user scope: %s", preferenceId)}
+	}
+
+	isPreferenceEnabledAtOrgScope := preference.IsEnabledForScope(OrgAllowedScope)
+	// get the value from the org scope if enabled at org scope
+	if isPreferenceEnabledAtOrgScope {
+		orgPreference := PreferenceKV{}
+
+		query := `SELECT preference_id , preference_value FROM org_preference WHERE preference_id=$1 AND org_id=$2;`
+
+		err := db.Get(&orgPreference, query, preferenceId, orgId)
+
+		// if there is error in getting values and its not an empty rows error return from here
+		if err != nil && err != sql.ErrNoRows {
+			return nil, &model.ApiError{Typ: model.ErrorExec, Err: fmt.Errorf("error in getting org preference values: %s", err.Error())}
+		}
+
+		// if there is no error update the preference value with value from org preference
+		if err == nil {
+			preferenceValue.PreferenceValue = orgPreference.PreferenceValue
+		}
+	}
+
+	// get the value from the user_preference table, if exists return this value else the one calculated in the above step
+	userPreference := PreferenceKV{}
+
+	query := `SELECT preference_id, preference_value FROM user_preference WHERE preference_id=$1 AND user_id=$2;`
+	err := db.Get(&userPreference, query, preferenceId, userId)
+
+	if err != nil && err != sql.ErrNoRows {
+		return nil, &model.ApiError{Typ: model.ErrorExec, Err: fmt.Errorf("error in getting user preference values: %s", err.Error())}
+	}
+
+	if err == nil {
+		preferenceValue.PreferenceValue = userPreference.PreferenceValue
+	}
+
+	return &PreferenceKV{
+		PreferenceId:    preferenceValue.PreferenceId,
+		PreferenceValue: preference.SanitizeValue(preferenceValue.PreferenceValue),
+	}, nil
+}
+
+func UpdateUserPreference(ctx context.Context, preferenceId string, preferenceValue interface{}, userId string) (*PreferenceKV, *model.ApiError) {
+	// check if the preference id is valid
+	preference, seen := preferenceMap[preferenceId]
+	if !seen {
+		return nil, &model.ApiError{Typ: model.ErrorBadData, Err: fmt.Errorf("no such preferenceId exists: %s", preferenceId)}
+	}
+
+	// check if the preference is enabled at user scope
+	isPreferenceEnabledAtUserScope := preference.IsEnabledForScope(UserAllowedScope)
+	if !isPreferenceEnabledAtUserScope {
+		return nil, &model.ApiError{Typ: model.ErrorForbidden, Err: fmt.Errorf("preference is not enabled at user scope: %s", preferenceId)}
+	}
+
+	err := preference.IsValidValue(preferenceValue)
+	if err != nil {
+		return nil, err
+	}
+	// update the user preference values
+	query := `INSERT INTO user_preference(preference_id,preference_value,user_id) VALUES($1,$2,$3)
+	ON CONFLICT(preference_id,user_id) DO
+	UPDATE SET preference_value= $2 WHERE preference_id=$1 AND user_id=$3;`
+
+	_, dberrr := db.Exec(query, preferenceId, preferenceValue, userId)
```

Comment:
I see `preference_value` is TEXT in db. Do you have some specific serialization format? For example, how to list type `preferenceValue` get stored in DB? You may not have that case right now but I want to understand how does the ser/deser work

Reply:
the values stored in the DB are stored as TEXT. as in this file we have `SanitizeValue` function which takes care of deserialising the values. similar to this when the need comes we will have the serialising function as well where we will perform the serialisations

Reply:
`TEXT` is the data type, not a serialization format. Are you saying we will write a custom ser/deser for preference values?

Reply:
yes when writing values for non standard types like list or objects we will serialise them to strings ( `stringify`) and write to DB and when the `sanitizeValue` function will take care of non serialising when returning the values